// cz-conventional-changelog-cmyr/dist/index.js
'use strict';

var wrap = require('word-wrap');
var map = require('lodash.map');
var longest = require('longest');
var chalk = require('chalk');
var conventionalCommitTypes = require('conventional-commit-types');
var commitizen = require('commitizen');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var wrap__default = /*#__PURE__*/_interopDefaultLegacy(wrap);
var map__default = /*#__PURE__*/_interopDefaultLegacy(map);
var longest__default = /*#__PURE__*/_interopDefaultLegacy(longest);
var chalk__default = /*#__PURE__*/_interopDefaultLegacy(chalk);
var conventionalCommitTypes__default = /*#__PURE__*/_interopDefaultLegacy(conventionalCommitTypes);

const config$1 = {
    type: {
        description: 'è¯·é€‰æ‹©æ‚¨è¦æäº¤çš„æ›´æ”¹ç±»åž‹',
        enum: {
            feat: {
                description: 'ä¸€ä¸ªæ–°åŠŸèƒ½(feature)',
                title: 'Features',
                emoji: 'âœ¨',
            },
            fix: {
                description: 'ä¸€ä¸ªé”™è¯¯ä¿®å¤(bug fix)',
                title: 'Bug Fixes',
                emoji: 'ðŸ›',
            },
            docs: {
                description: 'ä»…æ–‡æ¡£æ›´æ”¹',
                title: 'Documentation',
                emoji: 'ðŸ“š',
            },
            style: {
                description: 'ä¸å½±å“ä»£ç å«ä¹‰çš„ä»£ç æ ·å¼æ›´æ”¹(ç©ºæ ¼ã€æ ¼å¼ã€ç¼ºå°‘åˆ†å·ç­‰)',
                title: 'Styles',
                emoji: 'ðŸ’Ž',
            },
            refactor: {
                description: 'æ—¢ä¸ä¿®å¤é”™è¯¯ä¹Ÿä¸æ·»åŠ åŠŸèƒ½çš„ä»£ç æ›´æ”¹(ä»£ç é‡æž„)',
                title: 'Code Refactoring',
                emoji: 'ðŸ“¦',
            },
            perf: {
                description: 'æé«˜æ€§èƒ½çš„ä»£ç æ›´æ”¹',
                title: 'Performance Improvements',
                emoji: 'ðŸš€',
            },
            test: {
                description: 'æ·»åŠ ç¼ºå¤±çš„æµ‹è¯•æˆ–çº æ­£çŽ°æœ‰çš„æµ‹è¯•',
                title: 'Tests',
                emoji: 'ðŸš¨',
            },
            build: {
                description: 'å½±å“æž„å»ºç³»ç»Ÿæˆ–å¤–éƒ¨ä¾èµ–é¡¹çš„æ›´æ”¹ (ç¤ºä¾‹èŒƒå›´: gulp, broccoli, npm)',
                title: 'Builds',
                emoji: 'ðŸ› ',
            },
            ci: {
                description: 'å¯¹æˆ‘ä»¬çš„ CI é…ç½®æ–‡ä»¶å’Œè„šæœ¬çš„æ›´æ”¹ (ç¤ºä¾‹èŒƒå›´: Travis, Circle, BrowserStack, SauceLabs)',
                title: 'Continuous Integrations',
                emoji: 'âš™ï¸',
            },
            chore: {
                description: 'ä¸ä¿®æ”¹ src æˆ–æµ‹è¯•æ–‡ä»¶çš„å…¶ä»–æ›´æ”¹',
                title: 'Chores',
                emoji: 'â™»ï¸',
            },
            revert: {
                description: 'å›žé€€ä¹‹å‰çš„æäº¤',
                title: 'Reverts',
                emoji: 'ðŸ—‘',
            },
        },
    },
    scope: {
        description: 'æ­¤æ›´æ”¹çš„èŒƒå›´æ˜¯ä»€ä¹ˆ(ä¾‹å¦‚ç»„ä»¶æˆ–æ–‡ä»¶å): (æŒ‰ Enter é”®è·³è¿‡)',
    },
    subject: {
        description: 'å¯¹å˜åŒ–å†™ä¸€ä¸ªç®€çŸ­çš„ã€å‘½ä»¤å¼çš„æè¿°:\n',
    },
    body: {
        description: 'æä¾›æ›´è¯¦ç»†çš„æ›´æ”¹æè¿°: (æŒ‰ Enter é”®è·³è¿‡)\n',
    },
    isBreaking: {
        description: 'æ˜¯å¦æœ‰ä»»ä½•ç ´åæ€§å˜åŒ–(BREAKING CHANGE)?',
    },
    breakingBody: {
        description: 'ä¸€ä¸ª BREAKING CHANGE æäº¤éœ€è¦ä¸€ä¸ª bodyã€‚è¯·è¾“å…¥å¯¹æäº¤æœ¬èº«çš„æ›´é•¿æè¿°:\n',
    },
    breaking: {
        description: 'æè¿°ç ´åæ€§å˜åŒ–:\n',
    },
    isIssueAffected: {
        description: 'æ­¤æ›´æ”¹ä¼šå½±å“ä»»ä½•æœªè§£å†³çš„ issues å—?',
    },
    issuesBody: {
        description: 'å¦‚æžœ issues å·²å…³é—­ï¼Œåˆ™æäº¤éœ€è¦ä¸€ä¸ªä¸»ä½“ã€‚è¯·è¾“å…¥å¯¹æäº¤æœ¬èº«çš„æ›´é•¿æè¿°:\n',
    },
    issues: {
        description: 'æ·»åŠ é—®é¢˜å‚è€ƒ (ä¾‹å¦‚ "fix #123", "re #123".):\n',
    },
};

const filter = function (array) {
    return array.filter((x) => x);
};
const headerLength = function (answers) {
    return (answers.type.length + 2 + (answers.scope ? answers.scope.length + 2 : 0));
};
const maxSummaryLength = function (options, answers) {
    return options.maxHeaderWidth - headerLength(answers);
};
const filterSubject = function (subject, disableSubjectLowerCase) {
    subject = subject.trim();
    if (!disableSubjectLowerCase && subject.charAt(0).toLowerCase() !== subject.charAt(0)) {
        subject =
            subject.charAt(0).toLowerCase() + subject.slice(1, subject.length);
    }
    while (subject.endsWith('.')) {
        subject = subject.slice(0, subject.length - 1);
    }
    return subject;
};
// This can be any kind of SystemJS compatible module.
// We use Commonjs here, but ES6 or AMD would do just
// fine.
function engine (options) {
    const types = options.types;
    const length = longest__default["default"](Object.keys(types)).length + 1;
    const choices = map__default["default"](types, (type, key) => ({
        name: `${`${key}:`.padEnd(length)} ${type.description}`,
        value: key,
    }));
    return {
        // When a user runs `git cz`, prompter will
        // be executed. We pass you cz, which currently
        // is just an instance of inquirer.js. Using
        // this you can ask questions and get answers.
        //
        // The commit callback should be executed when
        // you're ready to send back a commit template
        // to git.
        //
        // By default, we'll de-indent your commit
        // template and will keep empty lines.
        prompter(cz, commit) {
            // Let's ask some questions of the user
            // so that we can populate our commit
            // template.
            //
            // See inquirer.js docs for specifics.
            // You can also opt to use another input
            // collection library if you prefer.
            cz.prompt([
                {
                    type: 'list',
                    name: 'type',
                    message: config$1.type.description,
                    choices,
                    default: options.defaultType,
                },
                {
                    type: 'input',
                    name: 'scope',
                    message: config$1.scope.description,
                    default: options.defaultScope,
                    filter(value) {
                        return options.disableScopeLowerCase
                            ? value.trim()
                            : value.trim().toLowerCase();
                    },
                },
                {
                    type: 'input',
                    name: 'subject',
                    message(answers) {
                        return (`${config$1.subject.description} (æœ€å¤š ${maxSummaryLength(options, answers)} ä¸ªå­—ç¬¦):\n`);
                    },
                    default: options.defaultSubject,
                    validate(subject, answers) {
                        const filteredSubject = filterSubject(subject, options.disableSubjectLowerCase);
                        // eslint-disable-next-line no-nested-ternary
                        return filteredSubject.length === 0
                            ? 'subject æ˜¯å¿…é¡»çš„'
                            : filteredSubject.length <= maxSummaryLength(options, answers)
                                ? true
                                : `subject é•¿åº¦å¿…é¡»å°äºŽæˆ–ç­‰äºŽ ${maxSummaryLength(options, answers)} ä¸ªå­—ç¬¦. å½“å‰é•¿åº¦ä¸º ${filteredSubject.length} ä¸ªå­—ç¬¦.`;
                    },
                    transformer(subject, answers) {
                        const filteredSubject = filterSubject(subject, options.disableSubjectLowerCase);
                        const color = filteredSubject.length <= maxSummaryLength(options, answers)
                            ? chalk__default["default"].green
                            : chalk__default["default"].red;
                        return color(`(${filteredSubject.length}) ${subject}`);
                    },
                    filter(subject) {
                        return filterSubject(subject, options.disableSubjectLowerCase);
                    },
                },
                {
                    type: 'input',
                    name: 'body',
                    message: config$1.body.description,
                    default: options.defaultBody,
                },
                {
                    type: 'confirm',
                    name: 'isBreaking',
                    message: config$1.isBreaking.description,
                    default: false,
                },
                {
                    type: 'input',
                    name: 'breakingBody',
                    default: '-',
                    message: config$1.breakingBody.description,
                    when(answers) {
                        return answers.isBreaking && !answers.body;
                    },
                    validate(breakingBody) {
                        return (breakingBody.trim().length > 0 ||
                            'BREAKING CHANGE å¿…é¡»è¦å¡«å†™ body!');
                    },
                },
                {
                    type: 'input',
                    name: 'breaking',
                    message: config$1.breaking.description,
                    when(answers) {
                        return answers.isBreaking;
                    },
                },
                {
                    type: 'confirm',
                    name: 'isIssueAffected',
                    message: config$1.isIssueAffected.description,
                    default: !!options.defaultIssues,
                },
                {
                    type: 'input',
                    name: 'issuesBody',
                    default: '-',
                    message: config$1.issuesBody.description,
                    when(answers) {
                        return (answers.isIssueAffected && !answers.body && !answers.breakingBody);
                    },
                },
                {
                    type: 'input',
                    name: 'issues',
                    message: config$1.issues.description,
                    when(answers) {
                        return answers.isIssueAffected;
                    },
                    default: options.defaultIssues ? options.defaultIssues : undefined,
                },
            ]).then((answers) => {
                const wrapOptions = {
                    trim: true,
                    cut: false,
                    newline: '\n',
                    indent: '',
                    width: options.maxLineWidth,
                };
                // parentheses are only needed when a scope is present
                const scope = answers.scope ? `(${answers.scope})` : '';
                // Hard limit this line in the validate
                const head = `${answers.type + scope}: ${answers.subject}`;
                // Wrap these lines at options.maxLineWidth characters
                const body = answers.body ? wrap__default["default"](answers.body, wrapOptions) : false;
                // Apply breaking change prefix, removing it if already present
                let breaking = answers.breaking ? answers.breaking.trim() : '';
                breaking = breaking
                    ? `BREAKING CHANGE: ${breaking.replace(/^BREAKING CHANGE: /, '')}`
                    : '';
                breaking = breaking ? wrap__default["default"](breaking, wrapOptions) : false;
                const issues = answers.issues ? wrap__default["default"](answers.issues, wrapOptions) : false;
                commit(filter([head, body, breaking, issues]).join('\n\n'));
            });
        },
    };
}

const config = commitizen.configLoader.load() || {};
const options = {
    types: config.types || config$1.type.enum || conventionalCommitTypes__default["default"].types,
    defaultType: process.env.CZ_TYPE || config.defaultType,
    defaultScope: process.env.CZ_SCOPE || config.defaultScope,
    defaultSubject: process.env.CZ_SUBJECT || config.defaultSubject,
    defaultBody: process.env.CZ_BODY || config.defaultBody,
    defaultIssues: process.env.CZ_ISSUES || config.defaultIssues,
    disableScopeLowerCase: process.env.DISABLE_SCOPE_LOWERCASE || config.disableScopeLowerCase,
    disableSubjectLowerCase: process.env.DISABLE_SUBJECT_LOWERCASE || config.disableSubjectLowerCase,
    maxHeaderWidth: process.env.CZ_MAX_HEADER_WIDTH &&
        parseInt(process.env.CZ_MAX_HEADER_WIDTH) ||
        config.maxHeaderWidth ||
        100,
    maxLineWidth: process.env.CZ_MAX_LINE_WIDTH &&
        parseInt(process.env.CZ_MAX_LINE_WIDTH) ||
        config.maxLineWidth ||
        100,
};
try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const commitlintLoad = require('@commitlint/load');
    commitlintLoad().then((clConfig) => {
        if (clConfig.rules) {
            const maxHeaderLengthRule = clConfig.rules['header-max-length'];
            if (typeof maxHeaderLengthRule === 'object' &&
                maxHeaderLengthRule.length >= 3 &&
                !process.env.CZ_MAX_HEADER_WIDTH &&
                !config.maxHeaderWidth) {
                options.maxHeaderWidth = maxHeaderLengthRule[2];
            }
        }
    });
}
catch (error) {
    //
}
module.exports = engine(options);
