// conventional-changelog-cmyr-config/writer-opts.js
'use strict'
const compareFunc = require('compare-func')
const Q = require('q')
const readFile = Q.denodeify(require('fs').readFile)
const resolve = require('path').resolve
const path = require('path')
const debug = require('debug')('conventional-changelog:cmyr-config"')
const _ = require('lodash')
// è‡ªå®šä¹‰é…ç½®
let pkgJson = {}
try {
    pkgJson = require(path.resolve(process.cwd(), './package.json'))
} catch (err) {
    console.error('no root package.json found')
}

const { changelog = {} } = pkgJson
const en = {
    feat: {
        title: 'âœ¨ Features',
        enable: true,
    },
    fix: {
        title: 'ðŸ› Bug Fixes',
        enable: true,
    },
    perf: {
        title: 'âš¡ Performance Improvements',
        enable: true,
    },
    revert: {
        title: 'âª Reverts',
        enable: true,
    },
    refactor: {
        title: 'â™» Code Refactoring',
        enable: true,
    },
    docs: {
        title: 'ðŸ“ Documentation',
        enable: false,
    },
    style: {
        title: 'ðŸ’„ Styles',
        enable: false,
    },
    test: {
        title: 'âœ… Tests',
        enable: false,
    },
    build: {
        title: 'ðŸ‘· Build System',
        enable: false,
    },
    ci: {
        title: 'ðŸ”§ Continuous Integration',
        enable: false,
    },
    chore: {
        title: 'ðŸŽ« Chores',
        enable: false,
    },
}
const zh = {
    feat: {
        title: 'âœ¨ æ–°åŠŸèƒ½',
        enable: true,
    },
    fix: {
        title: 'ðŸ› Bug ä¿®å¤',
        enable: true,
    },
    perf: {
        title: 'âš¡ æ€§èƒ½ä¼˜åŒ–',
        enable: true,
    },
    revert: {
        title: 'âª å›žé€€',
        enable: true,
    },
    refactor: {
        title: 'â™» ä»£ç é‡æž„',
        enable: true,
    },
    docs: {
        title: 'ðŸ“ æ–‡æ¡£',
        enable: false,
    },
    style: {
        title: 'ðŸ’„ é£Žæ ¼',
        enable: false,
    },
    test: {
        title: 'âœ… æµ‹è¯•',
        enable: false,
    },
    build: {
        title: 'ðŸ‘· æž„å»º',
        enable: false,
    },
    ci: {
        title: 'ðŸ”§ CI é…ç½®',
        enable: false,
    },
    chore: {
        title: 'ðŸŽ« å…¶ä»–æ›´æ–°',
        enable: false,
    },
}
const _settings = /(zh|cn|Han)/i.test(changelog && changelog.language) ? zh : en
const defaultOptions = {
    bugsUrl: false,
    authorName: false,
    authorEmail: false,
    settings: _settings,
}
const options = Object.assign({}, defaultOptions, changelog)
debug('options: %o', options)
const { bugsUrl, authorName, authorEmail } = options

changelog.settings = changelog.settings || {}

const settings = _.fromPairs(_.toPairs(_settings).map(([key, value]) => {
    if (!changelog.settings[key]) {
        return [key, value]
    }
    const newValue = { title: '', enable: false }
    if (typeof changelog.settings[key].title === 'string') {
        newValue.title = changelog.settings[key].title
    } else {
        newValue.title = value.title
    }

    if (typeof changelog.settings[key].enable === 'boolean') {
        newValue.enable = changelog.settings[key].enable
    } else {
        newValue.enable = value.enable
    }
    return [key, newValue]
}))

let gitUserInfo = ''
if (authorName && authorEmail) {
    gitUserInfo = 'by: **{{authorName}}** ({{authorEmail}})'
} else if (authorName) {
    gitUserInfo = 'by: **{{authorName}}**'
} else if (authorEmail) {
    gitUserInfo = 'by: ({{authorEmail}})'
}

module.exports = Q.all([
    readFile(resolve(__dirname, './templates/template.hbs'), 'utf-8'),
    readFile(resolve(__dirname, './templates/header.hbs'), 'utf-8'),
    readFile(resolve(__dirname, './templates/commit.hbs'), 'utf-8'),
    readFile(resolve(__dirname, './templates/footer.hbs'), 'utf-8'),
]).spread((template, header, commit, footer) => {
    const writerOpts = getWriterOpts()

    writerOpts.mainTemplate = template
    writerOpts.headerPartial = header
    // æ›¿æ¢ commit.hbs æ¨¡æ¿ä¸­çš„ gitUserInfo
    writerOpts.commitPartial = commit.replace(/{{gitUserInfo}}/g, gitUserInfo)
    writerOpts.footerPartial = footer

    return writerOpts
})
// å¿…é¡»é€‰é¡¹ 
const requiredOption = ['feat', 'fix', 'perf', 'revert', 'refactor']
// å¯é€‰é€‰é¡¹
const optionalOptions = ['docs', 'style', 'test', 'build', 'ci', 'chore']
function getWriterOpts() {
    return {
        transform: (commit, context) => {
            let discard = true
            const issues = []
            debug('commit ï¼š%s', commit)
            // if (commit.notes.length > 0) {
            //     discard = false
            // }
            commit.notes.forEach((note) => {
                note.title = 'ðŸ’¥ BREAKING CHANGES'
                discard = false
            })

            if (commit.revert) {
                commit.type = settings['revert'].title
            } else if (requiredOption.includes(commit.type)) {
                commit.type = settings[commit.type].title
            } else if (optionalOptions.includes(commit.type)) { // ä»¥ä¸Šä¸ºå¿…é¡»ï¼Œä»¥ä¸‹ä¸ºå¯é€‰
                if (!settings[commit.type].enable) {
                    commit.type = settings[commit.type].title
                    debug('è¯¥ commit ç±»åž‹ä¸ç”Ÿæˆæ—¥å¿—ï¼š%s', commit.type)
                    return
                }
                commit.type = settings[commit.type].title
            } else if (discard) {
                return
            } else {
                return
            }

            debug('commit.type ï¼š%s', commit.type)

            if (commit.scope === '*') {
                commit.scope = ''
            }

            if (typeof commit.hash === 'string') {
                commit.hash = commit.hash.substring(0, 7)
            }

            if (typeof commit.subject === 'string') {
                let url = context.repository
                    ? `${context.host}/${context.owner}/${context.repository}`
                    : context.repoUrl
                if (url) {
                    url = `${url}/issues/`
                    // Issue URLs.
                    commit.subject = commit.subject.replace(/#([0-9]+)/g, (_, issue) => {
                        issues.push(issue)
                        return `[#${issue}](${url}${issue})`
                    })
                }
                if (context.host) {
                    // User URLs.
                    commit.subject = commit.subject.replace(
                        /\B@([a-z0-9](?:-?[a-z0-9/]){0,38})/g,
                        (_, username) => {
                            if (username.includes('/')) {
                                return `@${username}`
                            }

                            return `[@${username}](${context.host}/${username})`
                        },
                    )
                }
            }

            // remove references that already appear in the subject
            commit.references = commit.references.filter((reference) => {
                if (issues.indexOf(reference.issue) === -1) {
                    return true
                }

                return false
            })

            if (bugsUrl) {
                commit.references = commit.references.map((ref) => ({
                    ...ref,
                    bugsUrl,
                }))
            }

            return commit
        },
        groupBy: 'type',
        commitGroupsSort: 'title',
        commitsSort: ['scope', 'subject'],
        noteGroupsSort: 'title',
        notesSort: compareFunc,
    }
}
